yt_idefix/data_structures.py

```diff
@@ -427,15 +427,199 @@ def _get_time(self):
 
     def _set_code_unit_attributes(self):
         """Conversion between physical units and code units."""
-        length_unit = self.parameters.get("unit_length", 1.0)
-        velocity_unit = self.parameters.get("unit_velocity", 1.0)
-        density_unit = self.parameters.get("unit_density", 1.0)
 
-        self.length_unit = self.quan(length_unit, "cm")
-        self.velocity_unit = self.quan(velocity_unit, "cm/s")
-        density_unit = self.quan(density_unit, "g/cm**3")
+        # Pluto's base units are length, velocity and density, but here we consider
+        # length, mass and time as base units. Since it can make us easy to calculate
+        # all units when self.units_override is not None.
+
+        # Default values of Pluto's base units which are stored in self.parameters
+        # if they can be read from definitions.h
+        # Otherwise, they are set to unity in cgs.
+        pluto_units = {
+            "length_unit": self.quan(self.parameters.get("length_unit", 1.0), "cm"),
+            "velocity_unit": self.quan(
+                self.parameters.get("velocity_unit", 1.0), "cm/s"
+            ),
+            "density_unit": self.quan(
+                self.parameters.get("density_unit", 1.0), "g/cm**3"
+            ),
+        }
+
+        uo_size = len(self.units_override)
+        if uo_size > 0 and uo_size < 3:
+            ytLogger.info(
+                "Less than 3 units were specified in units_override (got %s). "
+                "Need to rely on PLUTO's internal units to derive other units",
+                uo_size,
+            )
 
-        self.mass_unit = density_unit * self.length_unit ** 3
-        self.time_unit = self.length_unit / self.velocity_unit
-        self.magnetic_unit = np.sqrt(4.0 * np.pi * density_unit) * self.velocity_unit
+        uo_cache = self.units_override.copy()
+        while len(uo_cache) < 3:
+            unit, value = pluto_units.popitem()
+            # If any Pluto's base unit is specified in units_override, it'll be preserved
+            if unit in uo_cache:
+                continue
+            uo_cache[unit] = value
+            # Make sure the combination of units are able to derive base units
+            # If self.units_override is empty, skip this part
+            if uo_size > 0:
+                try:
+                    self._validate_units_override_keys(uo_cache)
+                    ytLogger.info("Relying on %s: %s.", unit, uo_cache[unit])
+                except ValueError:
+                    # It means the combination is invalid
+                    del uo_cache[unit]
+
+        # Some unit_setup_functions
+        # The condition statements are essential to prevent getting stuck in infinite recursion
+        # However they are fragile, one must be cautious to change them
+        def _setup_time_unit(cache):
```

Comment:
so it looks like we slightly misunderstood each other (no big deal !). I suggested to rename these functions as `_setup_<x>_unit` when they were designed around side effects on an outer-scope dictionary. Now that you evolved the design to return actual values instead, a more appropriate naming convention should be `_get_<x>_unit`. I'm sorry if my original comment caused confusion.

I am not opposing your original design btw. It's fine to use side effects if you want, especially in nested functions like these.

I'm thinking however this setup function, and the myriad of nested functions you defined, look more and more like a separate, tiny class with a very specific responsibility. What do you think ?

Reply:
Oh it’s my first time seeing the term “side effects” in programming, I thought you mean it's bad. Now I know its real meaning.

Due to their similar function names, I wanted to use `eval` instead of a dictionary here. But I remembered you said that we should avoid using `eval`.

> 'm thinking however this setup function, and the myriad of nested functions you defined, look more and more like a separate, tiny class with a very specific responsibility. What do you think ?

For me, it's like a `unyt.UnitSystem` which however only accepts mass, length, and time. So if you recommend, I'm willing to separate this part and wrap it into a class tomorrow. I’ve been thinking it's a little bit too much to define it in this method.

Reply:
Sure. I'm not familiar with how extensible unyt.UnitSystem is but if you think it's appropriate you could maybe subclass it. Otherwise just move most of this function into a (private) class and call it a day :)